Add doc page about `n_jobs` (#2768)
intel/scikit-learn-intelex: doc/sources/parallelism.rst

.. Copyright contributors to the oneDAL project

.. Licensed under the Apache License, Version 2.0 (the "License");
.. you may not use this file except in compliance with the License.
.. You may obtain a copy of the License at

..      http://www.apache.org/licenses/LICENSE-2.0

.. Unless required by applicable law or agreed to in writing, software
.. distributed under the License is distributed on an "AS IS" BASIS,
.. WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
.. See the License for the specific language governing permissions and
.. limitations under the License.

.. include:: substitutions.rst

#####################
Parallelism Specifics
#####################

|sklearnex| supports the :term:`n_jobs` parameter of the original |sklearn| with the following differences:

* `n_jobs` parameter is supported for all estimators patched by |sklearnex|,
  while |sklearn| enables it for selected estimators only.
* `n_jobs` estimator parameter sets the number of threads used by the underlying |oneDAL|.
* |sklearnex| doesn't use `joblib` for parallelism in patched estimators and functions.

> * The only low-level parallelism library used by |sklearnex| is oneTBB (through oneDAL and MKL).

Reviewer: ```suggestion
* The only low-level parallelism library used by |sklearnex| is [oneTBB](https://github.com/uxlfoundation/oneTBB) (through the |onedal| and [oneMKL](https://www.intel.com/content/www/us/en/developer/tools/oneapi/onemkl.html)).
```

* The `threading` parallel backend of `joblib` is not supported by |sklearnex|. User needs to specify desired number of threads 
  for each estimator directly using `n_jobs` parameter to avoid thread oversubscription.
* If `n_jobs` is not specified |sklearnex| uses all available threads whereas |sklearn| is single-threaded by default.

|sklearnex| follows the same rules as |sklearn| for

> `the calculation of the 'n_jobs' parameter value <https://scikit-learn.org/stable/glossary.html#term-n_jobs>`__.

Reviewer: ```suggestion
`the calculation of the :term:`n_jobs` parameter value.
```

|sklearnex| tries to determine the optimal number of threads per job using hints proded by `joblib`.
If ``n_jobs`` is not specified for underlying estimator(s), |sklearnex| sets it to the number of available threads
(usually the number of logical CPUs divided by `n_jobs` set for higher-level parallelized entities).

.. note::
    Environment variables such as `OMP_NUM_THREADS`, `MKL_NUM_THREADS`, `OPENBLAS_NUM_THREADS`, and others used by
    low-level parallelism libraries do not affect |sklearnex|.

.. note::
    ``n_jobs`` has no effect if computations are performed on GPU.

.. note::
    `threadpoolctl` context has no effect on |sklearnex| threading if `n_jobs` is specified and non-negative.
    If `n_jobs` is equal to `0` or not specified then the number from `threadpoolctl` is propagated to |sklearnex|.
    If `n_jobs` is negative then the `threadpoolctl`'s number will be `max(1, n_threadpoolctl + n_jobs + 1)`.

.. note::
    |sklearnex| threading doesn't automatically avoid nested parallelism when used in conjunction with OpenMP and/or with joblib or python threads.

To track the actual number of threads used by estimators from the |sklearnex|,

> set the `DEBUG` :ref:`verbosity setting <verbose>`.

Reviewer: I do not see any log with the number of threads when doing this.

Example:
```python
import os
os.environ["SKLEARNEX_VERBOSE"] = "DEBUG"
import numpy as np
from sklearnex.linear_model import Ridge
from sklearn.model_selection import GridSearchCV
rng = np.random.default_rng(seed=123)
X = rng.standard_normal(size=(100,10))
y = rng.standard_normal(X.shape[0])
Ridge().fit(X, y)
```
```
DEBUG:sklearnex: Assigned method '<host_backend>.linear_model.regression.train' to 'BaseLinearRegression.train'
DEBUG:sklearnex: Assigned method '<host_backend>.linear_model.regression.infer' to 'BaseLinearRegression.infer'
DEBUG:sklearnex: Assigned method '<host_backend>.linear_model.regression.model' to 'BaseLinearRegression.model'
DEBUG:sklearnex: Assigned method '<host_backend>.linear_model.regression.partial_train_result' to 'BaseIncrementalLinear.partial_train_result'
DEBUG:sklearnex: Assigned method '<host_backend>.linear_model.regression.partial_train' to 'BaseIncrementalLinear.partial_train'
DEBUG:sklearnex: Assigned method '<host_backend>.linear_model.regression.finalize_train' to 'BaseIncrementalLinear.finalize_train'
DEBUG:sklearnex: Assigned method '<host_backend>.logistic_regression.classification.train' to 'LogisticRegression.train'
DEBUG:sklearnex: Assigned method '<host_backend>.logistic_regression.classification.infer' to 'LogisticRegression.infer'
DEBUG:sklearnex: Assigned method '<host_backend>.logistic_regression.classification.model' to 'LogisticRegression.model'
INFO:sklearnex: sklearn.linear_model.Ridge.fit: running accelerated version on CPU
DEBUG:sklearnex: Dispatching function 'linear_model.regression.train' with policy <onedal._onedal_py_host.host_policy object at 0x7fb689e02770> to Backend(<module 'onedal._onedal_py_host' from '/home/dcortes/repos/scikit-learn-intelex/onedal/_onedal_py_host.cpython-311-x86_64-linux-gnu.so'>, is_dpc=False, is_spmd=False)
```

Reviewer: Thanks. Then the paragraph could be removed, since it doesn't work the way it describes. It'd nevertheless be helpful to have documentation about global settings to understand that message better.

|sklearnex| tries to determine the optimal number of threads per job using hints provided by :mod:`joblib` / ``threadpoolctl``..
If ``n_jobs`` is not specified for underlying estimator(s), |sklearnex| sets it to the number of available threads
(usually the number of logical CPUs divided by `n_jobs` set for higher-level parallelized entities).

.. note::
    Environment variables such as `OMP_NUM_THREADS`, `MKL_NUM_THREADS`, `OPENBLAS_NUM_THREADS`, and others used by
    low-level parallelism libraries do not affect |sklearnex|.

.. note::
    ``n_jobs`` has no effect if computations are performed on GPU.

.. note::
    `threadpoolctl` context has no effect on |sklearnex| threading if `n_jobs` is specified and non-negative.
    If `n_jobs` is equal to `0` or not specified then the number from `threadpoolctl` is propagated to |sklearnex|.
    If `n_jobs` is negative then the `threadpoolctl`'s number will be `max(1, n_threadpoolctl + n_jobs + 1)`.

.. note::
    |sklearnex| threading doesn't automatically avoid nested parallelism when used in conjunction with OpenMP and/or python threads.

To track the actual number of threads used by estimators from the |sklearnex|,

> set the `DEBUG` :ref:`verbosity setting <verbose>`. If |sklearnex|'s number of threads

Author: https://github.com/uxlfoundation/scikit-learn-intelex/pull/2768#discussion_r2560191296
does it answer this question as well?

> is different from the number of CPU cores it will be mentioned in logs.

Reviewer: I think the code checks for number of threads, not physical cores.

Reviewer: It looks like we might actually have different defaults depending on how oneDAL is called.

Nevertheless, as far as the python code is concerned, I see it takes the info from `cpu_count`, wich returns number of logical threads:
https://github.com/uxlfoundation/scikit-learn-intelex/blob/ebda062d8b40682e35dbd54215e3a7aad5cbd38a/daal4py/sklearn/_n_jobs_support.py#L102

..      http://www.apache.org/licenses/LICENSE-2.0

.. Unless required by applicable law or agreed to in writing, software
.. distributed under the License is distributed on an "AS IS" BASIS,
.. WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
.. See the License for the specific language governing permissions and
.. limitations under the License.

.. include:: substitutions.rst
.. _parallelism:

#####################
Parallelism Specifics
#####################

|sklearnex| supports the `n_jobs <https://scikit-learn.org/stable/glossary.html#term-n_jobs>`_ parameter
of the original |sklearn| with the following differences:

* `n_jobs` parameter is supported for all estimators patched by |sklearnex|,
  while |sklearn| enables it for selected estimators only.
* `n_jobs` estimator parameter sets the number of threads used by the underlying |oneDAL|.
* |sklearnex| doesn't use `joblib` for parallelism in patched estimators and functions.
* The only low-level parallelism library used by |sklearnex| is oneTBB (through oneDAL and MKL).
* The `threading` parallel backend of `joblib` is not supported by |sklearnex|. User needs to specify desired number of threads

>     for each estimator directly using `n_jobs` parameter to avoid thread oversubscription.

Reviewer: ```suggestion
  for each estimator directly using `n_jobs` parameter to avoid thread oversubscription.
```
For consistent spacing.

> * If `n_jobs` is not specified |sklearnex| uses all available threads while |sklearn| is single-threaded by default.

Reviewer: @Vika-F @Alexandr-Solovev Is it always the case that oneDAL will use the full number of threads in the system? Or does it for example drop low-power E cores in laptops?

Reviewer 2: By default it uses the full number of threads on all the hardware

> The only exception is multiclass LogisticRegression, which uses :external+joblib:doc:`joblib <index>` for parallelism across classes.

Reviewer: ```suggestion
The only exception is multiclass :obj:`sklearn.linear_model.LogisticRegression`, which uses :external+joblib:doc:`joblib <index>` for parallelism across classes.
```

|sklearnex| follows the same rules as |sklearn| for
`the calculation of the 'n_jobs' parameter value <https://scikit-learn.org/stable/glossary.html#term-n_jobs>`__.

> When Scikit-learn's utilities with built-in parallelism are used (for example, `GridSearchCV` or `VotingClassifier`),

Reviewer: ```suggestion
When Scikit-learn's utilities with built-in parallelism are used (for example, :obj:`sklearn.model_selection.GridSearchCV` or :obj:`sklearn.model_selection.VotingClassifier`),
```

|sklearnex| tries to determine the optimal number of threads per job using hints provided by `joblib`.
If ``n_jobs`` is not specified for underlying estimator(s), |sklearnex| sets it to the number of available threads
(usually the number of logical CPUs divided by `n_jobs` set for higher-level parallelized entities).

.. note::
    Environment variables such as `OMP_NUM_THREADS`, `MKL_NUM_THREADS`, `OPENBLAS_NUM_THREADS`, and others used by
    low-level parallelism libraries does not affect |sklearnex|.

.. note::
    ``n_jobs`` has no effect if computations are performed on GPU.

.. note::

>     `threadpoolctl` context has no effect on |sklearnex| threading if `n_jobs` is specified and non-negative.

Reviewer: Nevermind the original question - on a deeper look, the latest release of sklearn doesn't register itself to be configurable through threadpool.

But still, the paragraph currently says:
> `threadpoolctl` context has no effect on |sklearnex| threading if `n_jobs` is specified and non-negative.

So then: can `threadpoolctl` contexts somehow have any effect if `n_jobs` is either not specified or is negative?

>     `threadpoolctl` context has no effect on |sklearnex| threading if `n_jobs` is specified and non-negative.

Reviewer: Got it. Then it could be explicit about it and mention that threads from a threadpool context will be used if `n_jobs` is not specified or is zero. I get from that if-else that negative numbers would still override threadpool contexts though.

If ``n_jobs`` is not specified for underlying estimator(s), |sklearnex| sets it to the number of available threads
(usually the number of logical CPUs divided by `n_jobs` set for higher-level parallelized entities).

.. note::
    Environment variables such as `OMP_NUM_THREADS`, `MKL_NUM_THREADS`, `OPENBLAS_NUM_THREADS`, and others used by
    low-level parallelism libraries do not affect |sklearnex|.

.. note::
    ``n_jobs`` has no effect if computations are performed on GPU.

.. note::
    `threadpoolctl` context has no effect on |sklearnex| threading if `n_jobs` is specified and non-negative.
    If `n_jobs` is equal to `0` or not specified then the number from `threadpoolctl` is propagated to |sklearnex|.
    If `n_jobs` is negative then the `threadpoolctl`'s number will be `max(1, n_threadpoolctl + n_jobs + 1)`.

.. note::
    |sklearnex| threading doesn't automatically avoid nested parallelism when used in conjunction with OpenMP and/or python threads.

.. warning::
    If several instances of |sklearnex| algorithms are run in parallel with different ``n_jobs`` parameter it is not guaranteed that the

>     parameter would be propagated properly to each instance due to the known issue with [oneTBB](https://github.com/uxlfoundation/oneTBB).

Author: That's what I learned from the talk with @avolkov-intel

>     parameter would be propagated properly to each instance due to the known issue with [oneTBB](https://github.com/uxlfoundation/oneTBB).

Reviewer: Also perhaps it could clarify that this is about parallel python threads, not parallel processed which is more common in python.

>     parameter would be propagated properly to each instance due to the known issue with [oneTBB](https://github.com/uxlfoundation/oneTBB).

Author: Yes, I understand, I should have asked: was it ever observed?

.. note::
    `threadpoolctl` context has no effect on |sklearnex| threading if `n_jobs` is specified and non-negative.
    If `n_jobs` is equal to `0` or not specified then the number from `threadpoolctl` is propagated to |sklearnex|.
    If `n_jobs` is negative then the `threadpoolctl`'s number will be `max(1, n_threadpoolctl + n_jobs + 1)`.

.. note::
    |sklearnex| threading doesn't automatically avoid nested parallelism when used in conjunction with OpenMP and/or python threads.

.. warning::
    If several instances of |sklearnex| algorithms are run sequentially and the ``n_jobs`` parameter for the first run
    is significantly greater than for subsequent ones it may result in performance degradation due to a known issue
    with [oneTBB](https://github.com/uxlfoundation/oneTBB).

.. warning::
    If several instances of |sklearnex| algorithms are run in parallel with different ``n_jobs`` parameters
    it is not guaranteed that the parameter will be propagated correctly to each instance.

To track the actual number of threads used by estimators from the |sklearnex|,
set the `DEBUG` :ref:`verbosity setting <verbose>`. If |sklearnex|'s number of threads

> is different from the number of CPU cores it will be mentioned in logs.

Reviewer: ```suggestion
Setting the `DEBUG` :ref:`verbosity setting <verbose>` will produce logs
indicating when the number of threads used is different from the default
(number of logical threads in the machine).
```

* `n_jobs` estimator parameter sets the number of threads used by the underlying |onedal|.
* |sklearnex| doesn't use :mod:`joblib` for parallelism in patched estimators and functions.
* The only low-level parallelism library used by |sklearnex| is [oneTBB](https://github.com/uxlfoundation/oneTBB) 
  (through the |onedal| and [oneMKL](https://www.intel.com/content/www/us/en/developer/tools/oneapi/onemkl.html)).
* If `n_jobs` is not specified |sklearnex| uses all available threads whereas |sklearn| is single-threaded by default.

|sklearnex| follows the same rules as |sklearn| for
the calculation of the :term:`n_jobs` parameter value.

When |sklearn|'s utilities with built-in parallelism are used 
(for example, :obj:`sklearn.model_selection.GridSearchCV` or :obj:`sklearn.model_selection.VotingClassifier`),
|sklearnex| tries to determine the optimal number of threads per job using hints provided by :mod:`joblib` / ``threadpoolctl``..
If ``n_jobs`` is not specified for underlying estimator(s), |sklearnex| sets it to the number of available threads
(usually the number of logical CPUs divided by `n_jobs` set for higher-level parallelized entities).

.. note::

>     Environment variables such as `OMP_NUM_THREADS`, `MKL_NUM_THREADS`, `OPENBLAS_NUM_THREADS`, and others used by

Reviewer: Any ideas about whether [mkl_service](https://github.com/IntelPython/mkl-service) somehow would have an effect?

  (through the |onedal| and `oneMKL <https://www.intel.com/content/www/us/en/developer/tools/oneapi/onemkl.html>`__).
* If `n_jobs` is not specified, |sklearnex| uses all available threads whereas |sklearn| is single-threaded by default.
  Note that the deprecated :doc:`daal4py <daal4py>` module uses a global configuration instead of per-object ``n_jobs`` arguments, 
  with the default also being all available threads.

|sklearnex| follows the same rules as |sklearn| for
the calculation of the :term:`n_jobs` parameter value.

When |sklearn|'s utilities with built-in parallelism are used 
(for example, :obj:`sklearn.model_selection.GridSearchCV` or :obj:`sklearn.model_selection.VotingClassifier`),
|sklearnex| tries to determine the optimal number of threads per job using hints provided by :mod:`joblib` / ``threadpoolctl``..
If ``n_jobs`` is not specified for underlying estimator(s), |sklearnex| sets it to the number of available threads
(usually the number of logical CPUs divided by `n_jobs` set for higher-level parallelized entities).

.. note::
    Environment variables such as `OMP_NUM_THREADS`, `MKL_NUM_THREADS`, `OPENBLAS_NUM_THREADS`, and others used by

>     low-level parallelism libraries do not affect |sklearnex| as well as 

Reviewer: ```suggestion
    low-level parallelism libraries do not affect |sklearnex|, nor does the
```
